In [3]:
%run /Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebooks/candidate/TF_IDF/jobrec_tfidf_baseline.ipynb
%run /Users/nguyenduykhanh/Documents/GraduationProject/Recommendation_Project/recommendation/notebooks/candidate/TF_IDF/jobrec_tfidf_improved.ipynb


Jobs shape: (89277, 5)
CVs shape: (962, 2)
Jobs columns: ['title', 'description', 'company', 'location', 'source']
CVs columns: ['Category', 'Resume']


,title,description,company,location,source,cleaned_description,job_text
0,support technologist ii 2024-02216,**description and functions** ----------------...,State of Wyoming,"Rock Springs, WY",it,**description and functions** ----------------...,support technologist ii 2024-02216 **descripti...
1,computer technician,***if you would like to apply for this positio...,Medford Township Public Schools,"Medford, NJ",it,***if you would like to apply for this positio...,computer technician ***if you would like to ap...
2,it technician,i need a sped up recording converted to text. ...,Law Office of Ruben and Ruben LLC,,it,i need a sped up recording converted to text. ...,it technician i need a sped up recording conve...


,Category,Resume
0,Data Science,Skills * Programming Languages: Python (pandas...
1,Data Science,Education Details \r\nMay 2013 to May 2017 B.E...
2,Data Science,"Areas of Interest Deep Learning, Control Syste..."


job_texts: (89277,) cv_texts: (962,)
job_tfidf shape: (89277, 30000)
Demo CV index: 0


,title,company,location,semantic_score
72593,data analyst – nlp data science (bhjob22048_676),ITmPowered,"Ontario, CA",0.216367
73838,data scientist,American Savings Bank Hawaii,"Honolulu, HI, US",0.216030
76041,senior fraud data analyst,Newfold Digital,US,0.208217
74132,data scientist,POWER-tek Global Inc.,"White Plains, NY, US",0.199516
69162,data scientist,GHD,"Houston, TX, US",0.193592
72691,principal data scientist,octus,"New York, NY, US",0.192770
75637,associate fraud strategy data scientist,BILL,"San Jose, CA, US",0.190726
75941,associate fraud strategy data scientist,BILL,"San Jose, CA, USA",0.190473
76544,data science analyst (fraud strategy),Tiger Analytics,"Chicago, IL, US",0.189957
27622,fraud intelligence analyst,North.,"Remote, US",0.189423


Jobs shape: (89277, 5)
CVs shape: (962, 2)
Jobs columns: ['title', 'description', 'company', 'location', 'source']
CVs columns: ['Category', 'Resume']


,title,description,company,location,source,cleaned_description,job_text
0,support technologist ii 2024-02216,**description and functions** ----------------...,State of Wyoming,"Rock Springs, WY",it,**description and functions** ----------------...,support technologist ii 2024-02216 **descripti...
1,computer technician,***if you would like to apply for this positio...,Medford Township Public Schools,"Medford, NJ",it,***if you would like to apply for this positio...,computer technician ***if you would like to ap...
2,it technician,i need a sped up recording converted to text. ...,Law Office of Ruben and Ruben LLC,,it,i need a sped up recording converted to text. ...,it technician i need a sped up recording conve...


,Category,Resume
0,Data Science,Skills * Programming Languages: Python (pandas...
1,Data Science,Education Details \r\nMay 2013 to May 2017 B.E...
2,Data Science,"Areas of Interest Deep Learning, Control Syste..."


job_texts: (89277,) cv_texts: (962,)
job_tfidf shape: (89277, 30000)


,title,company,location,semantic_score,final_score,job_tax,job_sen,cv_tax,cv_sen,matched_skills,missing_skills
71116,data scientist,Brigham Young University,"Provo, UT, US",0.178472,0.434167,data_ai,mid,data_ai,senior,"[ai, cloud, deep learning, docker, git]","[automation, aws, azure, bi, etl]"
69267,lead data scientist,AXIANS,US,0.154859,0.409582,data_ai,senior,data_ai,senior,"[ai, angular, cloud, css, deep learning]","[api, aws, azure, bi, django]"
71376,director of data science & artificial intellig...,Fannie Mae,"Washington, DC, US",0.161941,0.392998,data_ai,senior,data_ai,senior,"[ai, cloud, deep learning, docker, git]","[api, automation, aws, azure, bi]"
76844,data scientist ii- assurance analysis,Honeywell,"Kansas City, MO, US",0.179117,0.382470,data_ai,mid,data_ai,senior,"[ai, cloud, deep learning, git, machine learning]","[bi, hadoop, pytorch, spark, tensorflow]"
77021,data science - senior - consumer products & re...,EY,"New York, NY, US",0.161634,0.382397,data_ai,senior,data_ai,senior,"[ai, cloud, express, java, javascript]","[api, aws, bi, pytorch, software engineer]"
75185,principal data scientist,Surescripts,"Minneapolis, MN, USA",0.157012,0.379624,data_ai,senior,data_ai,senior,"[ai, cloud, deep learning, express, machine le...","[api, aws, bi, gcp, pytorch]"
74967,principal data scientist,Surescripts,US,0.155722,0.378850,data_ai,senior,data_ai,senior,"[ai, cloud, deep learning, express, machine le...","[api, aws, bi, gcp, pytorch]"
74091,data scientist,DrFirst,US,0.155008,0.378421,data_ai,mid,data_ai,senior,"[ai, cloud, deep learning, git, machine learning]","[api, aws, azure, bi, gcp]"
69445,junior data scientist,Iron EagleX,"Fayetteville, NC, US",0.167328,0.375397,data_ai,junior,data_ai,senior,"[ai, deep learning, docker, git, machine learn...","[api, backend, bi, devops, etl]"
74602,data scientist,LP Building Solutions,"Nashville, TN, USA",0.182664,0.374182,data_ai,mid,data_ai,senior,"[ai, deep learning, java, javascript, machine ...","[bi, ios, nlp, warehouse]"


In [ ]:

# ===== Step 8: Evaluation metrics @K (baseline vs improved) =====
# topK job: bao nhiêu job đúng ngành so với CV
def _taxonomy_hit_at_k(df: pd.DataFrame) -> float:
    return float((df["job_tax"] == df["cv_tax"]).mean())

# topK job: bao nhiêu job lệch level so với CV 
def _seniority_mismatch_at_k(df: pd.DataFrame) -> float:
    return float((df["job_sen"] != df["cv_sen"]).mean())

# topK job: trùng bao nhiêu skill với CV
def _avg_skill_overlap_at_k(df: pd.DataFrame, cv_index: int) -> float:
    cv_sk = extract_skills(cv_texts_norm.iloc[cv_index])
    if not cv_sk:
        return 0.0
    overlaps = df["job_skills"].apply(lambda s: len(s & cv_sk) / max(1, len(cv_sk)))
    return float(overlaps.mean())

# matches có tồn tại không và bao nhiêu? 
def _explainability_rate_at_k(df: pd.DataFrame) -> float:
    return float(df["matched_skills"].apply(lambda x: isinstance(x, list) and len(x) > 0).mean())

def enrich_for_eval(df: pd.DataFrame, cv_index: int) -> pd.DataFrame:
    """Add job_tax/job_sen/job_skills + cv_tax/cv_sen + matched/missing for evaluation metrics."""
    out = df.copy()

    # ensure index maps to jobs row indices (this is true if df came from jobs.iloc[idx].copy())
    out["title"] = out["title"].fillna("").astype(str)

    # If cleaned_description not already in df, pull from global jobs by index
    if "cleaned_description" not in out.columns:
        out["cleaned_description"] = jobs.loc[out.index, "cleaned_description"].fillna("").astype(str)

    job_text_series = out["title"] + " " + out["cleaned_description"]

    # CV labels
    cv_text = cv_texts_norm.iloc[cv_index]
    out["cv_tax"] = cv_taxonomy(cv_index)
    out["cv_sen"] = infer_seniority_from_cv(cv_text)
    cv_sk = extract_skills(cv_text)

    # Job labels
    out["job_tax"] = out.apply(lambda r: classify_taxonomy_title_desc(r["title"], r["cleaned_description"]), axis=1)
    out["job_sen"] = out["title"].apply(infer_seniority_from_title)
    out["job_skills"] = job_text_series.apply(extract_skills)

    # Explain
    out["matched_skills"] = out["job_skills"].apply(lambda s: sorted(list(s & cv_sk))[:5] if isinstance(s, set) else [])
    out["missing_skills"] = out["job_skills"].apply(lambda s: sorted(list(s - cv_sk))[:5] if isinstance(s, set) else [])

    return out
 
def evaluate_option2_tfidf(sample_size: int = 50, top_k: int = 10, seed: int = 42):
    rng = np.random.default_rng(seed)
    n = len(resumes)
    idxs = rng.choice(n, size=min(sample_size, n), replace=False)

    rows = []
    for cv_idx in idxs:
        base = baseline_topk_tfidf(cv_idx, top_k=top_k)
        imp  = recommend_jobs_option2_tfidf(cv_idx, top_k=top_k)

        base2 = enrich_for_eval(base, cv_idx)
        imp2  = enrich_for_eval(imp,  cv_idx)

        rows.append({
            "type": "baseline",
            "cv_index": int(cv_idx),
            "TaxonomyHit@K": _taxonomy_hit_at_k(base2),
            "SeniorityMismatch@K": _seniority_mismatch_at_k(base2),
            "AvgSkillOverlap@K": _avg_skill_overlap_at_k(base2, cv_idx),
            "ExplainabilityRate@K": _explainability_rate_at_k(base2),
        })
        rows.append({
            "type": "improved",
            "cv_index": int(cv_idx),
            "TaxonomyHit@K": _taxonomy_hit_at_k(imp2),
            "SeniorityMismatch@K": _seniority_mismatch_at_k(imp2),
            "AvgSkillOverlap@K": _avg_skill_overlap_at_k(imp2, cv_idx),
            "ExplainabilityRate@K": _explainability_rate_at_k(imp2),
        })


    metrics_df = pd.DataFrame(rows)
    summary_df = metrics_df.groupby("type")[["TaxonomyHit@K","SeniorityMismatch@K","AvgSkillOverlap@K","ExplainabilityRate@K"]].mean()
    return metrics_df, summary_df

metrics_df, summary_df = evaluate_option2_tfidf(sample_size=50)
summary_df


,TaxonomyHit@K,SeniorityMismatch@K,AvgSkillOverlap@K,ExplainabilityRate@K
type,,,,
baseline,0.602,0.892,0.649855,0.986
improved,0.858,0.650,0.801691,1.000


In [5]:
# ===== Demo =====
metrics_df, summary_df = evaluate_option2_tfidf(sample_size=30, top_k=10, seed=42)
display(summary_df)
display(metrics_df.head(6))


,TaxonomyHit@K,SeniorityMismatch@K,AvgSkillOverlap@K,ExplainabilityRate@K
type,,,,
baseline,0.563333,0.943333,0.611728,0.98
improved,0.836667,0.626667,0.781048,1.00


,type,cv_index,TaxonomyHit@K,SeniorityMismatch@K,AvgSkillOverlap@K,ExplainabilityRate@K
0,baseline,616,0.0,1.0,0.191667,0.8
1,improved,616,0.6,0.7,0.633333,1.0
2,baseline,496,0.8,1.0,0.825000,1.0
3,improved,496,0.8,0.7,0.850000,1.0
4,baseline,405,0.0,1.0,0.430769,1.0
5,improved,405,0.4,0.8,0.538462,1.0


### Tổng kết kết quả đánh giá

| Chỉ số | Baseline | Đánh giá | Improved | Đánh giá |
|------|----------|---------|----------|---------|
| TaxonomyHit@K | ~0.58 | Chấp nhận được | ~0.85 | **Rất tốt** |
| SeniorityMismatch@K | ~0.92 | Rất kém | ~0.63 | **Trung bình – Tốt** |
| AvgSkillOverlap@K | ~0.63 | Hợp lý | ~0.80 | **Tốt** |
| ExplainabilityRate@K | ~0.98 | Tốt | 1.00 | **Đầy đủ** |


### Diễn giải TaxonomyHit@K
- Tỷ lệ các công việc trong top-K thuộc đúng ngành/ngữ nghĩa nghề nghiệp của hồ sơ ứng viên.

| Giá trị TaxonomyHit@K | Đánh giá |
|----------------------|---------|
| < 0.40 | Rất kém – Phần lớn job gợi ý lệch ngành |
| 0.40 – 0.60 | Mức chấp nhận được (baseline) |
| 0.60 – 0.75 | Tốt – Đa số job đúng ngành |
| **0.75 – 0.90** | **Rất tốt – Tính nhất quán ngành nghề cao** |
| > 0.90 | Xuất sắc – Gần như hoàn toàn đúng ngành |

- TopK(cv) = danh sách K job được gợi ý cho CV

- tax(job) = taxonomy của job

- tax(cv) = taxonomy của CV

TaxonomyHit@K
= (1 / K) * Σ I( tax(job_i) == tax(cv) )


### Diễn giải SeniorityMismatch@K (càng thấp càng tốt)

- Tỷ lệ các công việc trong top-K không phù hợp cấp độ kinh nghiệm so với hồ sơ ứng viên.

| Giá trị SeniorityMismatch@K | Đánh giá |
|----------------------------|---------|
| > 0.90 | Rất kém – Lệch cấp độ nghiêm trọng |
| 0.75 – 0.90 | Kém – Kiểm soát seniority chưa tốt |
| 0.60 – 0.75 | Trung bình – Vẫn còn lệch cấp độ |
| **0.40 – 0.60** | **Tốt – Phù hợp cấp độ tương đối** |
| < 0.40 | Rất tốt – Khớp cấp độ cao |

SeniorityMismatch@K
= (1 / K) * Σ I( sen(job_i) ≠ sen(cv) )


### Diễn giải AvgSkillOverlap@K

- Tỷ lệ trung bình các kỹ năng của ứng viên xuất hiện trong mô tả công việc.

| Giá trị AvgSkillOverlap@K | Đánh giá |
|--------------------------|---------|
| < 0.40 | Khớp kỹ năng yếu |
| 0.40 – 0.55 | Khớp kỹ năng hạn chế |
| 0.55 – 0.70 | Khớp kỹ năng ở mức hợp lý |
| **0.70 – 0.85** | **Khớp kỹ năng tốt** |
| > 0.85 | Khớp kỹ năng rất tốt |

- Skills(cv) = tập kỹ năng CV

- Skills(job) = tập kỹ năng job

SkillOverlap(job, cv)
= |Skills(job) ∩ Skills(cv)| / |Skills(cv)|

AvgSkillOverlap@K
= (1 / K) * Σ SkillOverlap(job_i, cv)


### Diễn giải ExplainabilityRate@K

- Tỷ lệ các công việc gợi ý có thể giải thích được lý do gợi ý (dựa trên taxonomy, seniority, kỹ năng).

| Giá trị ExplainabilityRate@K | Đánh giá |
|-----------------------------|---------|
| < 0.80 | Khả năng giải thích thấp |
| 0.80 – 0.90 | Giải thích một phần |
| 0.90 – 0.98 | Giải thích tốt |
| **≥ 0.98** | **Giải thích đầy đủ** |
